executor/join.go

```diff
@@ -286,34 +304,104 @@ func (e *HashJoinExec) fetchInnerRows(ctx context.Context, chkCh chan<- *chunk.C
 	}
 }
 
-// evalRadixBitNum evaluates the radix bit numbers.
-func (e *HashJoinExec) evalRadixBitNum() {
-	sv := e.ctx.GetSessionVars()
-	// Calculate the bit number needed when using radix partition.
-	if !sv.EnableRadixJoin {
-		return
+// partitionInnerRows re-order e.innerResults into sub-relations.
+func (e *HashJoinExec) partitionInnerRows() error {
+	if err := e.preAlloc4InnerParts(); err != nil {
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+	for i := 0; i < e.partConcurrency; i++ {
+		wg.Add(1)
+		workerID := i
+		go util.WithRecovery(func() {
+			e.doInnerPartition(workerID, wg)
+		}, nil)
+	}
+	return nil
+}
+
+func (e *HashJoinExec) doInnerPartition(startIdx int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	chkIdx, chkNum := startIdx, e.innerResult.NumChunks()
+	for ; chkIdx < chkNum; chkIdx += e.partConcurrency {
+		chk := e.innerResult.GetChunk(chkIdx)
+		for rowIdx, partPtr := range e.innerRowPrts[chkIdx] {
+			partIdx, rowPtr := partPtr.partitionIdx, partPtr.RowPtr
+			part := e.innerParts[partIdx]
+			part.Insert(rowPtr, chk.GetRow(rowIdx))
+		}
+	}
+}
+
+// preAlloc4InnerParts evaluates ptr4Partition for every inner row to help re-order
+// the inner relation.
+// TODO: we need to evaluate the skewness for the partitions size, if the
+// skewness exceeds a threshold, we do not use partition phase.
+func (e *HashJoinExec) preAlloc4InnerParts() (err error) {
```

Comment:
The time-consumption is about 1:2 ~ 1:1.5 for (preAlloc4InnerParts : doInnerPartition).

``` go
go tool pprof cpu.prof
Type: cpu
Time: Dec 13, 2018 at 7:58pm (CST)
Duration: 7.40s, Total samples = 12.29s (166.07%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) list partitionInnerRows
Total: 12.29s
ROUTINE ======================== github.com/pingcap/tidb/executor.(*HashJoinExec).partitionInnerRows in /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/join.go
         0      1.51s (flat, cum) 12.29% of Total
         .          .    306:}
         .          .    307:
         .          .    308:// partitionInnerRows re-order e.innerResults into sub-relations.
         .          .    309:func (e *HashJoinExec) partitionInnerRows() error {
         .          .    310:   t1 := time.Now()
         .      1.50s    311:   if err := e.preAlloc4InnerParts(); err != nil {
         .          .    312:           return err
         .          .    313:   }
         .       10ms    314:   logrus.Warning("PreAlloc:", time.Since(t1))
         .          .    315:   wg := &sync.WaitGroup{}
         .          .    316:   defer wg.Wait()
         .          .    317:   for i := 0; i < e.partConcurrency; i++ {
         .          .    318:           wg.Add(1)
         .          .    319:           workerID := i
ROUTINE ======================== github.com/pingcap/tidb/executor.(*HashJoinExec).partitionInnerRows.func1 in /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/join.go
         0      2.33s (flat, cum) 18.96% of Total
         .          .    316:   defer wg.Wait()
         .          .    317:   for i := 0; i < e.partConcurrency; i++ {
         .          .    318:           wg.Add(1)
         .          .    319:           workerID := i
         .          .    320:           go util.WithRecovery(func() {
         .      2.33s    321:                   e.doInnerPartition(workerID, wg)
         .          .    322:           }, nil)
         .          .    323:   }
         .          .    324:   return nil
         .          .    325:}
         .          .    326:
```

Reply:
What is the number of `e.partConcurrency` used in this benchmark?

Reply:
@XuHuaiyu I am still curious about the `e.partConcurrency` used here.

Reply:
@eurekaka 
It's 16
https://github.com/pingcap/tidb/pull/7911/files#diff-38079f9fe95df4d718f7d7b3689618aaR271